DOCS-4659 lib injection for hosts and containers (#16540)
DataDog/documentation: content/en/tracing/trace_collection/library_injection.md

## Requirements

- A recent [Datadog Agent v7][1] installation 

**Note**: Injection on arm64 is not supported. Injection with `musl` on Alpine Linux container images is untested.
## Install the preload library

1. Run the following commands, for example, with `apt`, where `<LANG>` is one of `java`, `nodejs`, `dotnet`, or `all`:

   ```sh
   sudo apt update
   sudo apt install -y datadog-apm-inject datadog-apm-library-<LANG>
   dd-install-ld-preload
   ```

2. The preload library injection only works in a new shell, so exit and open a new shell.

## Install the language and your app

1. For Java applications, ensure you have a JDK installed, for example:
   ```sh
   sudo apt install openjdk-17-jdk -y
   ```
   For NodeJS applications, ensure you have NodeJS installed, for example:
   ```sh
   curl -sL https://deb.nodesource.com/setup_16.x | sudo -E bash -
   sudo apt install nodejs -y
   ```
   For .NET applications, ensure you have the [.NET runtime installed][3].

2. If you haven't already, install your app.

## Configure the injection

> The following environment variables configure library injection. You can pass these in by `export` on the command line (`export DD_CONFIG_SOURCES=BASIC`), in your shell configuration, or by passing them in on the launch command:

Reviewer: ```suggestion
The following environment variables configure library injection. You can pass these in by `export` through the command line (`export DD_CONFIG_SOURCES=BASIC`), shell configuration, or launch command:
```

| `tracing_header_tags` | `dd.trace.header.tags` |    n/a    | `DD_TRACE_HEADER_TAGS` |
| `tracing_partial_flush_min_spans` | `dd.trace.partial.flush.min.spans` | `DD_TRACE_PARTIAL_FLUSH_MIN_SPANS` | `DD_TRACE_PARTIAL_FLUSH_ENABLED ` |
| `tracing_debug` | `dd.trace.debug` | `DD_TRACE_DEBUG` | `DD_TRACE_DEBUG` |
| `tracing_log_level` | `datadog.slf4j.simpleLogger.defaultLogLevel` | `DD_TRACE_LOG_LEVEL` |   n/a    |

Tracer library configuration options that aren't mentioned in the injection configuration are still available for use through properties or environment variables the usual way.

## Launch your services

Launch your services, indicating the preload library configuration in the launch command:

**Java app example**:
```sh
DD_CONFIG_SOURCES=BASIC java -jar <SERVICE_1>.jar &
DD_CONFIG_SOURCES=LOCAL:/etc/<SERVICE_2>/config.yaml;BASIC java -jar <SERVICE_2>.jar &
```

**Node app example**:
```sh
DD_CONFIG_SOURCES=BASIC node index.js &
DD_CONFIG_SOURCES=LOCAL:/etc/<SERVICE_2>/config.yaml;BASIC node index.js &
```

**.NET app example**:
```sh
DD_CONFIG_SOURCES=BASIC dotnet <SERVICE_1>.dll &
DD_CONFIG_SOURCES=LOCAL:/etc/<SERVICE_2>/config.yaml;BASIC dotnet <SERVICE_2>.dll &
```

> Exercise your application to start generating telemetry data, which you can see as [traces in APM][4].

Reviewer: ```suggestion
Execute your application to start generating telemetry data, which you can see as [traces in APM][4].
```

assuming "execute" is meant? unless exercising an application is also a thing

Author: @cswatt Yeah, exercise means "not just run it but also do some stuff in it that causes traces to happen."

[1]: https://app.datadoghq.com/account/settings#agent/overview
[2]: /tracing/trace_collection/library_config/
[3]: https://learn.microsoft.com/en-us/dotnet/core/install/linux-ubuntu
[4]: https://app.datadoghq.com/apm/traces
{{% /tab %}}

{{% tab "Agent on host, app in containers" %}}

<div class="alert alert-info">Tracing library injection on hosts and containers is in beta.</a></div>

> When your Agent is running on a host your services are running in containers, Datadog injects the tracing library by using a library that is preloaded and that overrides calls to `execve`, and a `runc` shim that intercepts container creation, and configures the initial process launched in a Docker container.

Reviewer: ```suggestion
When your Agent is running on a host, and your services are running in containers, Datadog injects the tracing library by using a preload library that overrides calls to `execve`. Datadog also uses a `runc` shim that intercepts container creation and configures the initial process launched in a Docker container.
```

Any newly started processes are intercepted and the specified instrumentation library is injected into the services.

## Requirements

- A recent [Datadog Agent v7][1] installation 
- [Docker Engine][2]

**Note**: Injection on arm64 is not supported. Injection with `musl` on Alpine Linux container images is untested.

## Install the preload library and shim

1. Run the following commands, for example, with `apt`, where `<LANG>` is one of `java`, `nodejs`, `dotnet`, or `all`:

   ```sh
   sudo apt update
   sudo apt install -y datadog-apm-inject datadog-apm-library-<LANG>
   dd-install-ld-preload
   dd-install-docker-shim
   ```

> 2. The preload library injection only works in a new shell, so exit and open a new shell.

Reviewer: ```suggestion
2. Exit and open a new shell to use the preload library.
```

| `tracing_partial_flush_min_spans` | `dd.trace.partial.flush.min.spans` | `DD_TRACE_PARTIAL_FLUSH_MIN_SPANS` | `DD_TRACE_PARTIAL_FLUSH_ENABLED ` |
| `tracing_debug` | `dd.trace.debug` | `DD_TRACE_DEBUG` | `DD_TRACE_DEBUG` |
| `tracing_log_level` | `datadog.slf4j.simpleLogger.defaultLogLevel` | `DD_TRACE_LOG_LEVEL` |   n/a    |

Tracer library configuration options that aren't mentioned in the injection configuration are still available for use through properties or environment variables the usual way.

## Specifying Unified Service Tags on containers

If the environment variables `DD_ENV`, `DD_SERVICE`, or `DD_VERSION` are specified in a service container image, those values are used to tag telemetry from the container.

If they are not specified, `DD_ENV` uses the `env` value set in the `/etc/datadog-agent/inject/docker_config.yaml` config file, if any. `DD_SERVICE` and `DD_VERSION` are derived from the name of the Docker image. An image with the name `my-service:1.0` is tagged with `DD_SERVICE` of `my-service` and a `DD_VERSION` of `1.0`.

## Launch your services

Start your Agent and launch your containerized services as usual.

> Exercise your application to start generating telemetry data, which you can see as [traces in APM][4].

Reviewer: ```suggestion
Execute your application to start generating telemetry data, which you can see as [traces in APM][4].
```

see previous note

[1]: https://app.datadoghq.com/account/settings#agent/overview
[2]: https://docs.docker.com/engine/install/ubuntu/
[3]: /tracing/trace_collection/library_config/
[4]: https://app.datadoghq.com/apm/traces

{{% /tab %}}

{{% tab "Containers" %}}

<div class="alert alert-info">Tracing library injection in containers is in beta.</a></div>

> When your Agent and services are running in Docker containers on the same host, Datadog injects the tracing library by using a library that is preloaded and that overrides calls to `execve`, and a `runc` shim that intercepts container creation, and configures the initial process launched in a Docker container.

Reviewer: ```suggestion
When your Agent and services are running in Docker containers on the same host, Datadog injects the tracing library by using a preload library that overrides calls to `execve`. Datadog also uses a `runc` shim that intercepts container creation and configures the initial process launched in a Docker container.
```

Any newly started processes are intercepted and the specified instrumentation library is injected into the services.

> # Requirements

Author: ```suggestion
## Requirements
```

- [Docker Engine][2]

**Note**: Injection on arm64 is not supported. Injection with `musl` on Alpine Linux container images is untested.

## Install the preload library and shim

1. Run the following commands, for example, with `apt`, where `<LANG>` is one of `java`, `nodejs`, `dotnet`, or `all`:

   ```sh
   sudo apt update
   sudo apt install -y datadog-apm-inject datadog-apm-library-<LANG>
   dd-install-ld-preload
   dd-install-docker-shim
   ```

> 2. The preload library injection only works in a new shell, so exit and open a new shell.

Reviewer: ```suggestion
2. Exit and open a new shell to use the preload library.
```

      - /opt/datadog/apm:/opt/datadog/apm
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /proc/:/host/proc/:ro
      - /sys/fs/cgroup/:/host/sys/fs/cgroup:ro
      - /sys/kernel/debug:/sys/kernel/debug
    cap_add:
      - SYS_ADMIN
      - SYS_RESOURCE
      - SYS_PTRACE
      - NET_ADMIN
      - NET_BROADCAST
      - NET_RAW
      - IPC_LOCK
      - CHOWN
    security_opt:
      - apparmor:unconfined
```
## Specifying Unified Service Tags on containers

If the environment variables `DD_ENV`, `DD_SERVICE`, or `DD_VERSION` are specified in a service container image, those values are used to tag telemetry from the container.

If they are not specified, `DD_ENV` uses the `env` value set in the `/etc/datadog-agent/inject/docker_config.yaml` config file, if any. `DD_SERVICE` and `DD_VERSION` are derived from the name of the Docker image. An image with the name `my-service:1.0` is tagged with `DD_SERVICE` of `my-service` and a `DD_VERSION` of `1.0`.

## Launch the Agent on docker

The `dd-agent` container must be launched before any service containers. Run:

```sh
docker-compose up -d dd-agent
```

## Launch your services

Launch your containerized services as usual.

> Exercise your application to start generating telemetry data, which you can see as [traces in APM][4].

Reviewer: ```suggestion
Execute your application to start generating telemetry data, which you can see as [traces in APM][4].
```

see previous notes
